### PR TITLE
feat: Mutation typed input

### LINKED ISSUE #2143: Mutation typed field parameters
The create and update operations currently use a JSON string to transmit document fields. Adding typed field parameters would enhance the user experience of DefraDB.

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2143 

## Description

This PR adds a typed input object for create and update mutations.

~~As a side effect of this change the relationship alias has been replaced with the `<type>_id` input field.~~

Relational sub-documents cannot be created from mutation input in this implementation.

Related SIP https://github.com/sourcenetwork/SIPs/discussions/10

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Make test

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: request/graphql/schema/generate.go [hunk 1]
```diff
@@ -517,6 +523,74 @@ func (g *Generator) buildTypes(
 	return objs, nil
 }
 
+// buildMutationInputTypes creates the input object types
+// for collection create and update mutation operations.
+func (g *Generator) buildMutationInputTypes(collections []client.CollectionDefinition) error {
+	for _, c := range collections {
+		// Copy the loop variable before usage within the loop or it
```
[the remark this answers is not in the document]
Comment (shahzadlone) [reply request/graphql/schema/generate.go:530]: Would appreciate a comment here mentioning `1.22` for regex search in the future, so can easily find some of these.

**Review by shahzadlone — CHANGES_REQUESTED** [review]
Why were the alias tests deleted? did we just drop support for aliases (i.e. a whole feature seems to have been dropped)?

**Comment by nasdf** [comment]
> > As a side effect of this change the relationship alias has been replaced with the _id input field.
> 
> I don't understand what is meant by 'replaced' with the `type_id`, because before we had both the 'type' alias and 'type_id'.
> 
> Why were the alias tests deleted? did we just drop support for aliases (i.e. a whole feature seems to have been dropped)?
> 
> IIRC this was a feature that was requested by one of the partners, and was implemented in:
> 
> * [feat: Add alias to `groupBy` related object #1579](https://github.com/sourcenetwork/defradb/pull/1579)
> * [feat: Allow relation alias on create and update #1609](https://github.com/sourcenetwork/defradb/pull/1609)
> 
> More context: https://discord.com/channels/427944769851752448/1061831110956421150/1113950104517365900

We talked about this briefly in the standup last week. Since the mutation input fields are now typed it is not possible to support relation field aliases (unless we add gql union types).

**Review by fredcarle — CHANGES_REQUESTED** [review]
In order to not get this PR hung up by the reverting of the aliasing feature, lets keep the aliasing for now and lets try to reach consensus separately on what we want to do about the possibility of allowing embedded foreign objects.

Without the removal of aliasing, I think this PR can me approved and merged.

**Review by fredcarle — APPROVED** [review]

**Review by fredcarle — APPROVED** [review]
LGTM! Awesome work! I love that we now have this implemented.

If you don't mind just adding the 1.22 reference as mentionned by Shahzad before merging.

**Review by shahzadlone — CHANGES_REQUESTED** [review]
Sorry just reviewing, thanks for adding alias stuff back. I am mid review will appreciate if possible to wait for me to finish.

File: planner/create.go [hunk 2]
```diff
@@ -135,24 +133,14 @@ func (n *createNode) Close() error {
...
-}
-
 // Explain method returns a map containing all attributes of this node that
 // are to be explained, subscribes / opts-in this node to be an explainablePlanNode.
 func (n *createNode) Explain(explainType request.ExplainType) (map[string]any, error) {
 	switch explainType {
 	case request.SimpleExplain:
-		return n.simpleExplain()
+		return map[string]any{
+			dataLabel: n.input,
```
Comment (shahzadlone) [inline planner/create.go:142]: todo: the name `dataLabel`doesn't make sense anymore. Please change to `inputLabel`

File: client/request/consts.go [hunk 3]
```diff
@@ -21,6 +21,7 @@ const (
 
 	Cid         = "cid"
 	Data        = "data"
```
Comment (shahzadlone) [inline client/request/consts.go:23]: todo: Please delete this `Data` variable now, it is now not used anywhere.

File: planner/update.go [hunk 4]
```diff
@@ -126,12 +131,7 @@ func (n *updateNode) simpleExplain() (map[string]any, error) {
 	}
 
 	// Add the attribute that represents the patch to update with.
-	data := map[string]any{}
-	err := json.Unmarshal([]byte(n.patch), &data)
-	if err != nil {
-		return nil, err
-	}
-	simpleExplainMap[dataLabel] = data
+	simpleExplainMap[dataLabel] = n.input
```
Comment (shahzadlone) [inline planner/update.go:134]: todo: Similar to other suggestions please change the variable name to be more relevant.

```suggestion
	simpleExplainMap[inputLabel] = n.input
```

**Review by shahzadlone — COMMENTED** [review]
Cheers for this PR! I like it a lot! Left some var renaming todos and in addition have this todo:

In `planner/explain.go` file we have on line 56: 
```
dataLabel           = "data"
```

After renaming the variable please don't forget to change the value from`"data"` to `"input"`.

OR if you want to use `request.Input` in all those instances and just remove the `dataLabel`/`inputLabel` , I am okay with that too :)

**Review by shahzadlone — APPROVED** [review]
Cheers for doing all the todos!

**Pull request merged**